extensions/html/server/src/modes/pathCompletion.ts

```diff
@@ -0,0 +1,118 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+'use strict';
+
+import { TextDocument, CompletionList, CompletionItemKind, CompletionItem } from 'vscode-languageserver-types';
+import { WorkspaceFolder } from 'vscode-languageserver-protocol/lib/protocol.workspaceFolders.proposed';
+import * as path from 'path';
+import * as fs from 'fs';
+import URI from 'vscode-uri';
+import { ICompletionParticipant } from 'vscode-html-languageservice/lib/htmlLanguageService';
+import { startsWith } from '../utils/strings';
+import { contains } from '../utils/arrays';
+
+export function getPathCompletionParticipant(
+	document: TextDocument,
+	workspaceFolders: WorkspaceFolder[] | undefined,
+	result: CompletionList
+): ICompletionParticipant {
+	return {
+		onHtmlAttributeValue: ({ tag, attribute, value, range }) => {
+
+			if (shouldDoPathCompletion(tag, attribute, value)) {
+				let workspaceRoot;
+
+				if (startsWith(value, '/')) {
+					if (!workspaceFolders || workspaceFolders.length === 0) {
+						return;
+					}
+
+					workspaceRoot = resolveWorkspaceRoot(document, workspaceFolders);
+				}
+
+				const suggestions = providePathSuggestions(value, URI.parse(document.uri).fsPath, workspaceRoot);
+				result.items = [...suggestions, ...result.items];
+			}
+		}
+	};
+}
+
+function shouldDoPathCompletion(tag: string, attr: string, value: string): boolean {
+	if (startsWith(value, 'http') || startsWith(value, 'https') || startsWith(value, '//')) {
+		return false;
+	}
+
+	if (PATH_TAG_AND_ATTR[tag]) {
+		if (typeof PATH_TAG_AND_ATTR[tag] === 'string') {
+			return PATH_TAG_AND_ATTR[tag] === attr;
+		} else {
+			return contains(<string[]>PATH_TAG_AND_ATTR[tag], attr);
+		}
+	}
+
+	return false;
+}
+
+export function providePathSuggestions(value: string, activeDocFsPath: string, root?: string): CompletionItem[] {
+	if (value.indexOf('/') === -1) {
+		return [];
+	}
+
+	if (startsWith(value, '/') && !root) {
+		return [];
+	}
+
+	const valueAfterLastSlash = value.slice(value.lastIndexOf('/') + 1);
+	const valueBeforeLastSlash = value.slice(0, value.lastIndexOf('/') + 1);
+	const parentDir = startsWith(value, '/')
+		? path.resolve(root, '.' + valueBeforeLastSlash)
+		: path.resolve(activeDocFsPath, '..', valueBeforeLastSlash);
+
+	return fs.readdirSync(parentDir).map(f => {
+		return {
+			label: f,
+			kind: isDir(path.resolve(parentDir, f)) ? CompletionItemKind.Folder : CompletionItemKind.File,
+			insertText: f.slice(valueAfterLastSlash.length)
```

Comment:
I don't understand why the 'slice' is here. Also, shouldn't there be a replace of the last segment after the slash?

Reply:
Instead of 'insertText' we should use 'textEdit' as to not rely on the work regexp.
(and to make the result testable)